### PR TITLE
Updated to Emscripten 1.39.16

### LINKED ISSUE #4803: Update to emscripten 1.39
We need to make it simpler to both bundle and the emscripten sdk.

Update with a package/build script for the sdk.
Update with install instructions

The big risk here is jumping to a major version without thorough vetting, but I'm confident we can all test the binary before releasing.

check list:

* [ ] package script - Darwin
* [ ] package script - Linux
* [ ] build.yml
* [ ] Extender server

### PR DESCRIPTION
**Pull request by JCash**
Added sdk package script
Added first embryo of using a local package path to build.py

### CODE REVIEW

File: README_EMSCRIPTEN.md [hunk 1]
```diff
@@ -81,14 +81,14 @@ corsproxy 172.16.11.23
...
+xhr.open(UTF8ToString(method), UTF8ToString(url), true);
 ```
 - and add
 ```
-var str_url = Pointer_stringify(url);
+var str_url = UTF8ToString(url);
 str_url = str_url.replace("http://", "http://172.16.11.23:9292/");
 str_url = str_url.replace("https://", "http://172.16.11.23:9292/");
-xhr.open(Pointer_stringify(method), str_url, true);
+xhr.open(UTF8ToString(method), str_url, true);
```
Comment (JCash) [inline README_EMSCRIPTEN.md:91]: Release notes: https://emscripten.org/docs/introducing_emscripten/release_notes.html

File: build_tools/waf_dynamo.py [hunk 2]
```diff
@@ -332,19 +332,23 @@ def default_flags(self):
...
         if 'wasm' == build_util.get_target_architecture():
             wasm_enabled = 1
             legacy_vm_support = 0
-            binaryen_method = "native-wasm"
+
+        emflags = ['WASM=%d' % wasm_enabled, 'LEGACY_VM_SUPPORT=%d' % legacy_vm_support, 'DISABLE_EXCEPTION_CATCHING=1', 'AGGRESSIVE_VARIABLE_ELIMINATION=1', 'PRECISE_F32=2',
+                   'EXTRA_EXPORTED_RUNTIME_METHODS=["stringToUTF8","ccall","stackTrace","UTF8ToString","callMain"]', 'EXPORTED_FUNCTIONS=["_main"]',
+                   'ERROR_ON_UNDEFINED_SYMBOLS=1', 'TOTAL_MEMORY=268435456']
+        emflags = zip(['-s'] * len(emflags), emflags)
+        emflags =[j for i in emflags for j in i]
```
Comment (JCash) [inline build_tools/waf_dynamo.py:343]: They don't clearly make a distinction between compile/link flags, and they're "safe to pass to both".

File: build_tools/waf_dynamo.py [hunk 3]
```diff
@@ -1206,7 +1210,7 @@ def linux_link_flags(self):
 @after('apply_obj_vars')
 def js_web_link_flags(self):
     platform = self.env['PLATFORM']
-    if 'web' in platform:
+    if 'web' in platform and 'test' in self.features:
```
Comment (JCash) [inline build_tools/waf_dynamo.py:1213]: Noticed that we added some code for headless mode, into all out executables.
This code is only for the tests.

We _do_ have one for the engine, but we don't seem to use it?

File: build_tools/waf_dynamo.py [hunk 4]
```diff
@@ -1472,7 +1476,7 @@ def detect(conf):
     remove_flag(conf.env['shlib_CXXFLAGS'], '-current_version', 1)
 
     # NOTE: We override after check_tool. Otherwise waf gets confused and CXX_NAME etc are missing..
-    if 'web' == build_util.get_target_os() and ('js' == build_util.get_target_architecture() or 'wasm' == build_util.get_target_architecture()):
+    if platform in ('js-web', 'wasm-web'):
```
Comment (JCash) [inline build_tools/waf_dynamo.py:1479]: Same code, more readable imho.

File: build_tools/waf_dynamo.py (hunk 2 above)
Comment (britzl) [reply build_tools/waf_dynamo.py:343]: This makes my head hurt "[j for i in emflags for j in i]" :-)

File: build_tools/waf_dynamo.py [hunk 5]
```diff
@@ -1540,7 +1548,7 @@ def detect(conf):
             conf.env['LUA_BYTECODE_ENABLE_32'] = 'yes'
 
     conf.env['STATICLIB_CARES'] = []
-    if platform != 'web':
+    if platform not in ['js-web', 'wasm-web']:
```
Comment (JCash) [inline build_tools/waf_dynamo.py:1551]: Previously, the compiler happily ignored the fact that the library didn't exist for that platform.

File: editor/tasks/leiningen/pack.clj [hunk 6]
```diff
@@ -26,7 +26,7 @@
   {"x86_64-darwin" {"bin" ["dmengine" "dmengine_release"]
                     "lib" ["libparticle_shared.dylib" "libtexc_shared.dylib"]}
    "x86-win32"     {"bin" ["dmengine.exe" "dmengine_release.exe"]
-                    "lib" ["particle_shared.dll"]}
```
Comment (JCash) [inline editor/tasks/leiningen/pack.clj:29]: Let's stop uploading artefacts for desktop platforms that we don't support.

File: engine/particle/src/wscript [hunk 7]
```diff
@@ -13,12 +13,16 @@ def build(bld):
...
                                 source = 'particle.cpp ../proto/particle/particle_ddf.proto')
 
-    particle_shared = bld.new_task_gen(features = 'cxx cshlib ddf',
-                                       includes = ['.', '../proto'],
-                                       target = 'particle_shared',
-                                       protoc_includes = '../proto',
-                                       uselib = 'DDF DLIB PLATFORM_SOCKET',
-                                       source = 'particle.cpp ../proto/particle/particle_ddf.proto')
+    # We only need this library in the editor
+    is_host = bld.env['PLATFORM'] in ('x86_64-linux', 'x86_64-win32', 'x86_64-darwin')
```
Comment (JCash) [inline engine/particle/src/wscript:17]: Let's stop build these for targets we don't need them for.

File: scripts/build.py [hunk 8]
```diff
@@ -554,37 +555,52 @@ def download_sdk(url, targetfolder, strip_components=1):
...
+    def get_ems_dir(self):
+        return join(self.ext, 'SDKs', 'emsdk-' + EMSCRIPTEN_VERSION_STR)
+
     def _form_ems_path(self):
-        path = join(self.ext, EMSCRIPTEN_DIR)
-        return path
+        upstream = join(self.get_ems_dir(), 'upstream', 'emscripten')
+        if os.path.exists(upstream):
+            return upstream
+        return join(self.get_ems_dir(), 'fastcomp', 'emscripten')
```
Comment (JCash) [inline scripts/build.py:566]: Depending on how the package was installed (with or without upstream/fastcomp), this code will detect it automatically. It should help anyone wanting to compile with any one of them.

File: scripts/build.py [hunk 9]
```diff
@@ -554,37 +555,52 @@ def download_sdk(url, targetfolder, strip_components=1):
...
-            print("Error. Could not download %s" % url)
-            sys.exit(1)
-        self._extract(dlpath, self.ext)
-        self.activate_ems()
+        # TODO: should eventually be moved to install_sdk
+        emsDir = self.get_ems_dir()
+
         os.environ['EMSCRIPTEN'] = self._form_ems_path()
+        os.environ['EM_CONFIG'] = join(self.get_ems_dir(), '.emscripten')
+        os.environ['EM_CACHE'] = join(self.get_ems_dir(), 'emscripten_cache')
```
Comment (JCash) [inline scripts/build.py:574]: Now _all_ emscripten files are located within `tmp/dynamo_home/ext/SDKs/emsdk-1.39.16` and are easy to remove and reinstall.

File: scripts/build.py [hunk 10]
```diff
@@ -554,37 +555,52 @@ def download_sdk(url, targetfolder, strip_components=1):
...
 
     def activate_ems(self):
+        version = EMSCRIPTEN_VERSION_STR
+        if 'fastcomp' in self._form_ems_path():
+            version += "-fastcomp"
+        run.env_command(self._form_env(), [join(self.get_ems_dir(), 'emsdk'), 'activate', version, '--embedded'])
+
+        # prewarm the cache
+        # Although this method might be more "correct", it also takes 10 minutes more than we'd like on CI
+        #run.env_command(self._form_env(), ['%s/embuilder.py' % self._form_ems_path(), 'build', 'SYSTEM', 'MINIMAL'])
```
Comment (JCash) [inline scripts/build.py:596]: The "correct" way took 10 minutes, and we don't want that on the CI.

**Review by britzl — APPROVED** [review]

File: scripts/build.py [hunk 11]
```diff
@@ -846,13 +862,12 @@ def archive_engine(self):
...
             self.upload_file(resources, '%s/android-resources.zip' % (full_archive_path))
 
-        libs = ['particle']
         if self.is_desktop_target():
-            libs.append('texc')
-        for lib in libs:
-            lib_name = format_lib('%s_shared' % (lib), self.target_platform)
-            lib_path = join(dynamo_home, 'lib', lib_dir, lib_name)
-            self.upload_file(lib_path, '%s/%s' % (full_archive_path, lib_name))
+            libs = ['texc', 'particle']
```
Comment (JCash) [inline scripts/build.py:866]: Not sure why "particle" shared library was never considered a "desktop" feature?

File: scripts/docker/Dockerfile [hunk 12]
```diff
@@ -69,6 +69,8 @@ RUN \
 
 RUN apt-get autoremove
 
+ENV LC_ALL C.UTF-8
```
Comment (JCash) [inline scripts/docker/Dockerfile:72]: Helped a little bit, but couldn't manage to build with emscripten using our linux docker.
More investigation needed.

File: scripts/package/package_emscripten.sh [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+VERSION=1.39.16
+URL=https://github.com/emscripten-core/emsdk/archive/${VERSION}.tar.gz
+PLATFORM=`uname`
+PLATFORM="$(tr [A-Z] [a-z] <<< "$PLATFORM")"
+
+PWD=`pwd`
+
+TARGET_PATH=${PWD}/local_sdks
+TMP=${TARGET_PATH}/_tmpdir
+TARGET=$TARGET_PATH/emsdk-${VERSION}-${PLATFORM}.tar.gz
```
Comment (JCash) [inline scripts/package/package_emscripten.sh:26]: Sadly, no windows support, yet.
It's not archived that way.
Perfect for a contribution, I think.

File: share/extender/build.yml [hunk 14]
```diff
@@ -300,20 +300,20 @@ platforms:
...
         libPaths:   ["{{dynamo_home}}/lib/js-web", "{{dynamo_home}}/ext/lib/js-web"]
         defines:    ["DM_PLATFORM_HTML5", "GL_ES_VERSION_2_0"]
-        flags:      ["-fno-exceptions"]
-        linkFlags:  ["--pre-js {{dynamo_home}}/share/js-web-pre.js", "--emit-symbol-map", "--llvm-lto", "0", "--memory-init-file", "0"]
-        emscriptenFlags: ["PRECISE_F32=2", "AGGRESSIVE_VARIABLE_ELIMINATION=1", "TOTAL_MEMORY=268435456", "DISABLE_EXCEPTION_CATCHING=1", "LEGACY_VM_SUPPORT=1", "WASM=0", "BINARYEN_METHOD=\"asmjs\"", "EXTRA_EXPORTED_RUNTIME_METHODS=[\"writeStringToMemory\",\"writeArrayToMemory\",\"stringToUTF8\",\"ccall\"]", "EXPORTED_FUNCTIONS=[\"_JSWriteDump\",\"_main\"]", "ERROR_ON_UNDEFINED_SYMBOLS=1"]
+        flags:      ["-fno-exceptions", "-fno-rtti", "-fPIC"]
+        linkFlags:  ["--emit-symbol-map", "--memory-init-file", "0", "-lidbfs.js"]
```
Comment (JCash) [inline share/extender/build.yml:315]: The IDBFS isn't linked by default anymore.

**Pull request merged**